pkg/models/iam/im/im.go

```diff
@@ -170,6 +206,101 @@ func (im *defaultIMOperator) CreateUser(user *iamv1alpha2.User) (*iamv1alpha2.Us
 	return user, nil
 }
 
+func (im *defaultIMOperator) authRateLimitEnabled() bool {
+	if im.authenticateRateLimiterMaxTries <= 0 || im.authenticateRateLimiterDuration == 0 {
+		return false
+	}
+	return true
+}
+
+func (im *defaultIMOperator) authRateLimitExceeded(user *iamv1alpha2.User) bool {
+	if !im.authRateLimitEnabled() {
+		return false
+	}
+	count := 0
+	for _, condition := range user.Status.Conditions {
+		if condition.Status == iamv1alpha2.ConditionTrue &&
+			condition.Type == iamv1alpha2.LoginFailure &&
+			time.Now().Sub(condition.LastTransitionTime.Time) <= im.authenticateRateLimiterDuration {
+			count++
+		}
+	}
+	return count >= im.authenticateRateLimiterMaxTries
+}
+
+func (im *defaultIMOperator) authFailRecord(user *iamv1alpha2.User, err error) {
+	if !im.authRateLimitEnabled() {
+		return
+	}
+	user = user.DeepCopy()
+	condition := iamv1alpha2.UserCondition{
+		Type:               iamv1alpha2.LoginFailure,
+		Status:             iamv1alpha2.ConditionTrue,
+		LastTransitionTime: metav1.Time{Time: time.Now()},
+		Reason:             "Authentication Failed",
+		Message:            err.Error(),
+	}
+	if user.Status.Conditions == nil {
+		user.Status.Conditions = make([]iamv1alpha2.UserCondition, 0)
+	}
+	user.Status.Conditions = append(user.Status.Conditions, condition)
+	// remove out date and out range record
+	record := 0
+	for i := len(user.Status.Conditions); i > 0; i-- {
```

Comment:
If you just want to remove condition out of slice, it's easy to understand if create another slice,  and copy only conditions needed over.

Reply:
fixed

Reply:
I don't think it's appropriate to save login attempts in the conditions. It's better to create another object, say Login, and logs every login attempt as a Login object. If we save login attempts as conditions, it's not easy to proceed, and user objects can be very big. For example, what if set maxFailureLoginAttempts to 1000.

Reply:
I think login record is also a part of user condition, we can easily get the user's status, it is necessary to set a reasonable maxFailureLoginAttempts, just like the resource limit needs a reasonable value. If maxFailureLoginAttempts is not needed, it can be set to zero. Even if maxFailureLoginAttempts is large than 1000, the user list response will not output `user.status.Conditions`. Tell the user how to set a reasonable value is better than adding a CRD.

Reply:
Why is here no user.status.Conditions in response?

Reply:
`user.status.Conditions` is redundant in the user list. Set `user.status.Conditions` to nil before the API(`/kapis/ iam.kubesphere.io/v1alpha2/users`) returns.

Reply:
What you did here is just filter response returned to console, the status is still there in the kubernetes API response